allenact/utils/inference.py

```diff
@@ -0,0 +1,122 @@
+from typing import Optional, cast, Tuple
+
+import attr
+import torch
+
+from allenact.algorithms.onpolicy_sync.policy import ActorCriticModel
+from allenact.algorithms.onpolicy_sync.storage import RolloutStorage
+from allenact.base_abstractions.experiment_config import ExperimentConfig, MachineParams
+from allenact.base_abstractions.misc import (
+    Memory,
+    ObservationType,
+    ActorCriticOutput,
+    DistributionType,
+)
+from allenact.base_abstractions.preprocessor import SensorPreprocessorGraph
+from allenact.utils.tensor_utils import batch_observations
+from allenact.utils import spaces_utils as su
+
+
+@attr.s(kw_only=True)
+class InferenceAgent:
+    actor_critic: ActorCriticModel = attr.ib()
+    rollout_storage: RolloutStorage = attr.ib()
+    device: torch.device = attr.ib()
+    sensor_preprocessor_graph: Optional[SensorPreprocessorGraph] = attr.ib()
+    steps_before_rollout_refresh: int = attr.ib(default=128)
+    memory: Optional[Memory] = attr.ib(default=None)
+    steps_taken_in_task: int = attr.ib(default=0)
+    last_action_flat: Optional = attr.ib(default=None)
+
+    def __attrs_post_init__(self):
+        self.actor_critic.eval()
+        self.actor_critic.to(device=self.device)
+        if self.memory is not None:
+            self.memory.to(device=self.device)
+        self.sensor_preprocessor_graph.to(self.device)
+
+        self.rollout_storage.to(self.device)
+        self.rollout_storage.set_partition(index=0, num_parts=1)
+
+    @classmethod
+    def from_experiment_config(
```

Comment:
Yah, basically the idea here is to make it easier for people to run inference with a model they have trained without explicitly running things through the allenact runner/engine. Here's a little script I threw together to test this out locally:
```python
import torch

from allenact.utils.inference import InferenceAgent
from projects.objectnav_baselines.experiments.robothor.objectnav_robothor_rgbd_resnet18gru_ddppo import (
    ObjectNavRoboThorRGBPPOExperimentConfig,
)

if __name__ == "__main__":
    exp_config = ObjectNavRoboThorRGBPPOExperimentConfig()
    agent = InferenceAgent.from_experiment_config(
        exp_config=ObjectNavRoboThorRGBPPOExperimentConfig(),
        device=torch.device("cpu"),
        checkpoint_path="pretrained_model_ckpts/robothor-objectnav-challenge-2021/Objectnav-RoboTHOR-RGBD-ResNetGRU-DDPPO/2021-02-09_22-35-15/exp_Objectnav-RoboTHOR-RGBD-ResNetGRU-DDPPO_0.2.0a_300M__stage_00__steps_000170207237.pt",
    )

    task_sampler = exp_config.make_sampler_fn(
        **exp_config.valid_task_sampler_args(process_ind=0, total_processes=1)
    )

    for _ in range(10):
        agent.reset()

        task = task_sampler.next_task()
        observations = task.get_observations()

        while not task.is_done():
            action = agent.act(observations=observations)
            observations = task.step(action).observation

        print({k: v for k, v in task.metrics().items() if k != "task_info"})
```
It would probably be a good idea to add a test for this...

Reply:
Really cool! I agree it would be nice to have a small test.

Reply:
Added (and it found a bug where experiment configs without sensor preprocessor graphs would break).